### PR TITLE
Add support for upsert memo

### PR DESCRIPTION
**Pull request by Quinn-With-Two-Ns**
Add support for upsert memo to the Java SDK and test server.

closes https://github.com/temporalio/sdk-java/issues/623

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java [hunk 1]
```diff
@@ -156,6 +156,11 @@ public void upsertTypedSearchAttributes(SearchAttributeUpdate<?>... searchAttrib
     next.upsertTypedSearchAttributes(searchAttributeUpdates);
   }
 
+  @Override
+  public void upsertMemo(Map<String, Object> memo) {
+    next.upsertSearchAttributes(memo);
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java:161]:
```suggestion
    next.upsertMemo(memo);
```

File: temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java [hunk 2]
```diff
@@ -1186,6 +1186,16 @@ public static void upsertTypedSearchAttributes(
     WorkflowInternal.upsertTypedSearchAttributes(searchAttributeUpdates);
   }
 
+  /**
+   * Updates a Workflow Memos by applying {@code memoUpdates} to the existing Memos set attached to
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:1190]: Can you clarify how a memo value is removed? Is this done by setting a null object? So can memo contain a null value for a key? In .NET we accept [memo update objects](https://dotnet.temporal.io/api/Temporalio.Workflows.Workflow.html#Temporalio_Workflows_Workflow_UpsertMemo_Temporalio_Workflows_MemoUpdate___) to remove any ambiguity. I forget what Go does here.

Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:1190]: Yes by setting a null value same as Go.

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java (hunk 1 above)
Comment (dandavison) [reply temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java:161]: Tests passed so looks like we must be missing some test coverage?

File: temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java (hunk 2 above)
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:1190]: Is there a difference between a null value with key present and an absent key? Specifically, if I do an `upsertMemo(Collections.singletonMap("key-that-exists-from-start", null))`, what will the raw memo be if I do describe workflow?

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java [hunk 3]
```diff
@@ -156,6 +156,11 @@ public void upsertTypedSearchAttributes(SearchAttributeUpdate<?>... searchAttrib
     next.upsertTypedSearchAttributes(searchAttributeUpdates);
   }
 
+  @Override
+  public void upsertMemo(Map<String, Object> memo) {
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java:160]: Does Java offer the ability for me to see the whole memo collection inside the workflow? For instance if I wanted to know all memo keys in a workflow, can I? Do we need to add that?

Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java:160]: No and No, at least no one has asked for it and I see no reason this PR would need to add it.

File: temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java (hunk 2 above)
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:1190]: `key-that-exists-from-start` would be removed by the server just like Go

Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/workflow/Workflow.java:1190]: :+1: May be worth mentioning in docs that that is how to remove a key

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java (hunk 3 above)
Comment (cretz) [reply temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java:160]: Works for me. My concern is if/when we do we need to make sure to more explicitly make sure null means remove from a `Map` POV.

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java (hunk 1 above)
Comment (Quinn-With-Two-Ns) [reply temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptorBase.java]: Yep added some test coverage for this, will need to think long term if there is some way I can test all the interceptors programatically

**Review by cretz — APPROVED** [review]

**Pull request merged**